src/Validator/Rules/KnownTypeNames.php

```diff
@@ -6,46 +6,89 @@
-use GraphQL\Language\Visitor;
-use GraphQL\Language\VisitorOperation;
+use GraphQL\Language\AST\TypeDefinitionNode;
+use GraphQL\Language\AST\TypeSystemDefinitionNode;
+use GraphQL\Language\AST\TypeSystemExtensionNode;
+use GraphQL\Type\Definition\Type;
 use GraphQL\Utils\Utils;
 use GraphQL\Validator\QueryValidationContext;
+use GraphQL\Validator\SDLValidationContext;
+use GraphQL\Validator\ValidationContext;
+use function in_array;
 
 /**
  * Known type names.
  *
  * A GraphQL document is only valid if referenced types (specifically
  * variable definitions and fragment conditions) are defined by the type schema.
+ *
+ * @phpstan-import-type VisitorArray from \GraphQL\Language\Visitor
  */
 class KnownTypeNames extends ValidationRule
 {
     public function getVisitor(QueryValidationContext $context): array
     {
-        $skip = static function (): VisitorOperation {
-            return Visitor::skipNode();
-        };
+        return $this->getASTVisitor($context);
+    }
+
+    public function getSDLVisitor(SDLValidationContext $context): array
+    {
+        return $this->getASTVisitor($context);
+    }
+
+    /**
+     * @phpstan-return VisitorArray
+     */
+    public function getASTVisitor(ValidationContext $context): array
+    {
+        /** @var array<int, string> $definedTypes */
+        $definedTypes = [];
+        foreach ($context->getDocument()->definitions as $def) {
+            if ($def instanceof TypeDefinitionNode) {
+                $definedTypes[] = $def->name->value;
+            }
+        }
+
+        $standardTypeNames = array_keys(Type::getAllBuiltInTypes());
 
         return [
-            // TODO: when validating IDL, re-enable these. Experimental version does not
-            // add unreferenced types, resulting in false-positive errors. Squelched
-            // errors for now.
-            NodeKind::OBJECT_TYPE_DEFINITION => $skip,
-            NodeKind::INTERFACE_TYPE_DEFINITION => $skip,
-            NodeKind::UNION_TYPE_DEFINITION => $skip,
-            NodeKind::INPUT_OBJECT_TYPE_DEFINITION => $skip,
-            NodeKind::NAMED_TYPE => static function (NamedTypeNode $node) use ($context): void {
-                $schema = $context->getSchema();
+            NodeKind::NAMED_TYPE => static function (NamedTypeNode $node, $_1, $parent, $_2, $ancestors) use ($context, $definedTypes, $standardTypeNames): void {
                 $typeName = $node->name->value;
-                $type = $schema->getType($typeName);
-                if ($type !== null) {
+                $schema = $context->getSchema();
+
+                if (in_array($typeName, $definedTypes, true)) {
+                    return;
+                }
+
+                if ($schema !== null && $schema->hasType($typeName)) {
+                    return;
+                }
+
+                $definitionNode = $ancestors[2] ?? $parent;
+                $isSDL = $definitionNode !== null && self::isSDLNode($definitionNode);
```

Comment:
```suggestion
                $isSDL = $definitionNode instanceof TypeSystemDefinitionNode || $definitionNode instanceof TypeSystemExtensionNode
```

Should be sufficient and more performant.